src/hooks/transactions/useEstimateFee.ts

```diff
@@ -0,0 +1,68 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges UI.
+//
+// Parity Bridges UI is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Parity Bridges UI is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges UI.  If not, see <http://www.gnu.org/licenses/>.
+
+import { useCallback } from 'react';
+import { useApiCallsContext } from '../../contexts/ApiCallsContextProvider';
+import { useSourceTarget } from '../../contexts/SourceTargetContextProvider';
+import { TransactionActionCreators } from '../../actions/transactionActions';
+import { useUpdateTransactionContext, useTransactionContext } from '../../contexts/TransactionContext';
+import { getSubstrateDynamicNames } from '../../util/getSubstrateDynamicNames';
+import useLaneId from '../chain/useLaneId';
+
+export const useEstimateFee = () => {
+  const { createType, stateCall } = useApiCallsContext();
+  const laneId = useLaneId();
+  const {
+    sourceChainDetails: { chain: sourceChain },
+    targetChainDetails: { chain: targetChain }
+  } = useSourceTarget();
+  const { dispatchTransaction } = useUpdateTransactionContext();
+  const { estimatedFeeMethodName } = getSubstrateDynamicNames(targetChain);
+  const { payload } = useTransactionContext();
+
+  const getEstimateFeeCall = useCallback(async () => {
+    if (!payload) {
+      return;
+    }
+    try {
+      dispatchTransaction(TransactionActionCreators.setCalculatingFee(true));
+      // Ignoring custom types missed for TS for now.
+      // Need to apply: https://polkadot.js.org/docs/api/start/typescript.user
+      // @ts-ignore
+      const payloadType = createType(sourceChain, 'OutboundPayload', payload);
+      // @ts-ignore
+      const messageFeeType = createType(sourceChain, 'MessageFeeData', {
+        lane_id: laneId,
+        payload: payloadType.toHex()
+      });
+
+      const estimatedFeeCall = await stateCall(sourceChain, estimatedFeeMethodName, messageFeeType.toHex());
+
+      // @ts-ignore
+      const estimatedFeeType = createType(sourceChain, 'Option<Balance>', estimatedFeeCall);
+      const estimatedFee = estimatedFeeType.toString();
+      dispatchTransaction(TransactionActionCreators.setEstimateFee(estimatedFee));
+    } catch (error) {
+      dispatchTransaction(TransactionActionCreators.setTransactionError(error, 'esstimateFee'));
+    } finally {
+      dispatchTransaction(TransactionActionCreators.setCalculatingFee(false));
+    }
```

Comment:
I really tried to make the `useApiGenericCall` hook to handle reducers dispatchers calls as you propose but in the end i end up with something like:

```
const setGlobalIsLoading = useCallback( (value)=> dispatchTransaction(Action.setIsLoading(value)), [dispatchTransaction] );

useApiGenericCall(call, {setGlobalIsLoading});
```

I see a couple of issues with this:

1) Adds more boilerplate since we have to wrap dispatchTransaction in useCallback. Without useCallback, the hook gets infinitive re render because of the different functions reference in each render. Also we would face this issue with using a local state of a context.  Using `dispatch` functions have a great benefit that is forgetting about the functions references in renders , so you avoid the useCallback.

2) the useApiGenericCall adds additional re renderings during the initial rendering because of it's initial state.

3) Makes useApiGenericCall quite messy imho.

I have shared more information here: https://gist.github.com/tomusdrw/6be7adb067ba44920931d8146f7467f7#gistcomment-3777056

Reply:
Regarding this file i opted for removing it from this PR since it was not used and it was created more as a boilerplate. 
Nikos in his draft PR #190 is actually using it so i prefer using his PR for including this function.

@wirednkod please be aware of this disscussion.

Reply:
> Adds more boilerplate

Wait, how removing the entire async code handling vs adding one `useCallback` is more boilerplate?

>  Without useCallback, the hook gets infinitive re render because of the different functions reference in each render.

Which hook? I don't see any `useEffect` here that could be triggered.

> Using dispatch functions have a great benefit that is forgetting about the functions references in renders , so you avoid the useCallback.

I'm not sure I understand this point. You need to pass `dispatch` to every `useEffect` anyway?

> the useApiGenericCall adds additional re renderings during the initial rendering because of it's initial state.

Well, the we can add initial state as a parameter? Also I don't think optimising re-renderings is a good metric currently. We are tackling more severe issues here like subscription leaks and incorrect async code handling (for instance using old values). If someone looses significant amount of money due to incorrect transaction being sent from the UI I'm pretty sure they would be convinced that it's because we wanted to avoid few re-renderings.

> Makes useApiGenericCall quite messy imho.

Yes, that's a fair point. We can consider having a separate hook-based async handler and dispatch-based async handler then? It's still an improvement because we only have 2 places in code which deal with async code instead of having it duplicated in random hooks/components.

Reply:
> Yes, that's a fair point. We can consider having a separate hook-based async handler and dispatch-based async handler then? It's still an improvement because we only have 2 places in code which deal with async code instead of having it duplicated in random hooks/components.

All your comments are valid , but i feel that it would be more effective to discuss on a code rather than explaining "if i do this, then happen this". I will try to create the equivalent useApiGenericCall for dispatchers then